File: frontend/src/services/trip.js
import api from "./api";

export function normalizeTrip(trip) {
  return {
    ...trip,
    vehicle_id: Number(trip.vehicle_id),
    driver_id: Number(trip.driver_id),
    cargo_weight: Number(trip.cargo_weight || 0),
    planned_distance: Number(trip.planned_distance || 0),
    actual_distance: Number(trip.actual_distance || 0),
    fuel_used: Number(trip.fuel_used || 0),
    revenue: Number(trip.revenue || 0),
  };
}

export function toTripPayload(values) {
  return {
    trip_number: values.trip_number,
    vehicle_id: Number(values.vehicle_id),
    driver_id: Number(values.driver_id),
    source: values.source,
    destination: values.destination,
    cargo_weight: String(values.cargo_weight || 0),
    planned_distance: String(values.planned_distance || 0),
    actual_distance: String(values.actual_distance || 0),
    fuel_used: String(values.fuel_used || 0),
    revenue: String(values.revenue || 0),
    status: values.status || "Draft",
    departure_time: values.departure_time || null,
    arrival_time: values.arrival_time || null,
    created_by: values.created_by || null,
  };
}

export async function listTrips(params = {}) {
  const { data } = await api.get("/trips/", { params });
  return data.map(normalizeTrip);
}

export async function createTrip(values) {
  const { data } = await api.post("/trips/", toTripPayload(values));
  return normalizeTrip(data);
}

export async function updateTrip(id, values) {
  const { data } = await api.put(`/trips/${id}`, toTripPayload(values));
  return normalizeTrip(data);
}

export async function deleteTrip(id) {
  await api.delete(`/trips/${id}`);
}

File: backend/app/models/trip.py
from sqlalchemy import (
    Column,
    DateTime,
    Enum,
    ForeignKey,
    Integer,
    Numeric,
    String
)
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func

from app.core.database import Base
from sqlalchemy import CheckConstraint, Index
from app.models.enums import TripStatus


class Trip(Base):
    __tablename__ = "trips"

    id = Column(Integer, primary_key=True, index=True)

    trip_number = Column(
        String(30),
        unique=True,
        nullable=False
    )

    vehicle_id = Column(
        Integer,
        ForeignKey("vehicles.id"),
        nullable=False
    )

    driver_id = Column(
        Integer,
        ForeignKey("drivers.id"),
        nullable=False
    )

    source = Column(
        String(150),
        nullable=False
    )

    destination = Column(
        String(150),
        nullable=False
    )

    cargo_weight = Column(
        Numeric(10, 2),
        nullable=False
    )

    planned_distance = Column(
        Numeric(10, 2),
        nullable=False
    )

    actual_distance = Column(
        Numeric(10, 2),
        default=0
    )

    fuel_used = Column(
        Numeric(10, 2),
        default=0
    )

    revenue = Column(
        Numeric(12, 2),
        default=0
    )

    status = Column(
        Enum(TripStatus),
        default=TripStatus.DRAFT,
        nullable=False
    )

    departure_time = Column(DateTime)

    arrival_time = Column(DateTime)

    created_by = Column(
        Integer,
        ForeignKey("users.id")
    )

    created_at = Column(
        DateTime(timezone=True),
        server_default=func.now()
    )

    updated_at = Column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now()
    )

    vehicle = relationship(
        "Vehicle",
        back_populates="trips"
    )

    driver = relationship(
        "Driver",
        back_populates="trips"
    )

    fuel_logs = relationship(
        "FuelLog",
        back_populates="trip"
    )

    expenses = relationship(
        "Expense",
        back_populates="trip"
    )
    
    __table_args__ = (

        CheckConstraint(
            "cargo_weight > 0",
            name="ck_trip_weight"
        ),

        CheckConstraint(
            "planned_distance > 0",
            name="ck_trip_distance"
        ),

        CheckConstraint(
            "actual_distance >= 0",
            name="ck_actual_distance"
        ),

        CheckConstraint(
            "fuel_used >= 0",
            name="ck_trip_fuel"
        ),

        CheckConstraint(
            "revenue >= 0",
            name="ck_trip_revenue"
        ),

        Index("idx_trip_vehicle", "vehicle_id"),
        Index("idx_trip_driver", "driver_id"),
        Index("idx_trip_status", "status"),
    )

    def __repr__(self):
        return f"<Trip {self.trip_number}>"